src/System.Composition.Hosting/src/System/Composition/Hosting/Core/CompositionOperation.cs

```diff
@@ -0,0 +1,121 @@
+    public sealed class CompositionOperation : IDisposable
+    {
+        private List<Action> _nonPrerequisiteActions;
+        private List<Action> _postCompositionActions;
+        private object _sharingLock;
+
+        // Construct using Run() method.
+        private CompositionOperation() { }
+
+        /// <summary>
+        /// Execute a new composition operation starting within the specified lifetime
+        /// context, for the specified activator.
+        /// </summary>
+        /// <param name="outermostLifetimeContext">Context in which to begin the operation (the operation can flow
+        /// to the parents of the context if requried).</param>
+        /// <param name="compositionRootActivator">Activator that will drive the operation.</param>
+        /// <returns>The composed object graph.</returns>
+        public static object Run(LifetimeContext outermostLifetimeContext, CompositeActivator compositionRootActivator)
+        {
+            Requires.ArgumentNotNull(outermostLifetimeContext, "outermostLifetimeContext");
+            Requires.ArgumentNotNull(compositionRootActivator, "compositionRootActivator");
+
+            using (var operation = new CompositionOperation())
+            {
+                var result = compositionRootActivator(outermostLifetimeContext, operation);
+                operation.Complete();
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Called during the activation process to specify an action that can run after all
+        /// prerequesite part dependencies have been satisfied.
+        /// </summary>
+        /// <param name="action">Action to run.</param>
+        public void AddNonPrerequisiteAction(Action action)
+        {
+            if (action == null)
+                throw new ArgumentNullException("action");
+
+            if (_nonPrerequisiteActions == null)
+                _nonPrerequisiteActions = new List<Action>();
+
+            _nonPrerequisiteActions.Add(action);
+        }
+
+        /// <summary>
+        /// Called during the activation process to specify an action that must run only after
+        /// all composition has completed. See OnImportsSatisfiedAttribute.
+        /// </summary>
+        /// <param name="action">Action to run.</param>
+        public void AddPostCompositionAction(Action action)
+        {
+            Requires.ArgumentNotNull(action, "action");
+
+            if (_postCompositionActions == null)
+                _postCompositionActions = new List<Action>();
+
+            _postCompositionActions.Add(action);
+        }
+
+        internal void EnterSharingLock(object sharingLock)
+        {
+            Assumes.NotNull(sharingLock, "Sharing lock is required");
+
+            if (_sharingLock == null)
+            {
+                _sharingLock = sharingLock;
+                Monitor.Enter(sharingLock);
```

Comment:
Thanks for the additional details.  I'm still not following 100%.  To be clear, the same CompositionOperation instance will never be used concurrently, e.g. the same instance will never be passed to GetOrCreate at the same time on two different threads?  If so, why is the Monitor.Enter necessary at all?  The way this is constructed, if EnterSharingLock has already been called, then _sharingLock will forever be non-null, so any subsequent callers won't even invoke Monitor.Enter, so what's the purpose of the Enter call?